### ISSUE TITLE
Arity in rewrite rules

### ISSUE THREAD

**Issue by francoisthire** [rlepigre/lambdapi#55]
I don't really understand how the arity restriction of lambdapi works. For example in the following signature:

```
A : Type.

def h : (A -> A) -> (A -> A) -> A -> A.
```

The following rule is accepted:

```
[aa] h aa (x => aa x) --> aa.
```

But not this one:

```
[aa] h (x => aa x) aa --> aa.
```

with the error message: "Arity mismatch for metavariable [aa].".

**Comment by rlepigre** [comment]
Note that this syntax for reduction rules is only provided for compatibility with Dedukti. You should rather use the following, where "metavariables" must always be fully applied.

```
h (x => ?aa[x]) (x => ?aa[x]) --> x:A => ?aa[x]
```

Your first rule is somehow desugared in such a way that `aa` is a metavariable of arity 0, and thus syntactic matching is used (not a Miller pattern). In the second rule is however desugared in a different way, and `aa` is made a metavariable of arity 1. Probably because it is used with arity 1 the first time it is encountered.

This is probably a buggy behaviour, and I'll investigate. In that particular case, we probably want to take as arity the minimal number of arguments to which `aa` is applied (that is, `0` in both cases). Alternatively, we can "eta-expand" as I did in the above rule.

**Comment by fblanqui** [comment]
We should take the minimum.

**Comment by fblanqui** [comment]
Should be fixed by 92e0cf5106e373d328844a5240544d9ea84ea205.
